File: apps/golang/oracle-tac/cmd/ingest.go
/*
Copyright © 2022 Andrea Lauri <andrea.lauri86@gmail.com>

Command ``ingest`` is used store new records within the db.

It can repeat the same operation in loop or a defined numbers of times. It can
be possible to define a delay between one operation and the next one and also
after how many operations commit the changes.
*/
package cmd

import (
	"fmt"
	"time"
	"io/ioutil"
	"strings"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

// ingestCmd represents the ingest command
var ingestCmd = &cobra.Command{
	Use:   "ingest",
	Short: "Insert new records into the table",
	Long:  `Insert new records into the table`,
	Run: func(cmd *cobra.Command, args []string) {
		iters, _ := cmd.Flags().GetInt("iters")
		delay, _ := cmd.Flags().GetFloat64("delay")
		commit_every, _ := cmd.Flags().GetInt("commit-every")

		// Define query parameters
		table := viper.GetViper().GetString("database.tableraw")
		filename := viper.GetViper().GetString("ingest.dumpfile")

		// Open the file and check error
		data, err := ioutil.ReadFile(filename)
		if err != nil {
			// TODO: replace panic command
			panic(err)
		}
		lines := strings.Split(string(data), "\n")

		step := 1
		for _, line := range lines[:iters] {
			// Exit condition
			if step > iters {
				break
			}

			// Prepare query with updated conditions
			query := fmt.Sprintf("INSERT INTO %s(year,track,data) VALUES(%s)", table, line)

			// Perform the query
			// db.DoQuery(conn, query)
			fmt.Fprintln(cmd.OutOrStdout(),
				fmt.Sprintf("[%d/%d] - %s", step, iters, query))

			// Commit changes
			if step%commit_every == 0 {
				// TODO: insert a database commit operation here
				fmt.Fprintln(cmd.OutOrStdout(),
					fmt.Sprintf("[%d/%d] - COMMIT", step, iters))
			}

			step += 1
			time.Sleep(time.Duration(delay) * time.Second)
		}

		// Chech the last commit
		if iters%commit_every != 0 {
			// TODO: insert a database commit operation here
			fmt.Fprintln(cmd.OutOrStdout(),
				fmt.Sprintf("[%d/%d] - COMMIT", iters, iters))
		}
	},
}

func init() {
	ingestCmd.Flags().Int("iters", 1, "repeat the same operation a given number of times")
	ingestCmd.Flags().Float64("delay", 0.25, "time to wait before the next iteration")
	ingestCmd.Flags().Int("commit-every", 1, "after how many operations perform a commit")
}

File: apps/python/otac/cmd/ingest.py
#!/usr/bin/python3


"""Command ``ingest`` is used store new records within the db.

It can repeat the same operation in loop or a defined numbers of times. It can
be possible to define a delay between one operation and the next one and also
after how many operations commit the changes.
"""


import time
import click

import cx_Oracle


@click.command()
@click.option('--iters',
              type=int,
              default=1,
              help='repeat the same operation a given number of times')
@click.option('--delay',
              type=float,
              default=0.25,
              help='time to wait before the next iteration')
@click.option('--commit-every',
              type=int,
              default=1,
              help='after how many operations perform a commit')
@click.pass_context
def ingest(ctx,
           iters: int,
           delay: float,
           commit_every: int) -> None:
    """Insert new records within the table"""

    # Instrumentation: Set up current module (ACCHK_REPORT)
    ctx.obj.conn.module = "otac.cmd.ingest"

    # Define query parameters
    data = open(ctx.obj.conf["ingest"]["dumpfile"]).readlines()
    try:
        for step, line in enumerate(data[:iters]):
            step += 1

            # Instrumentation: Set up module action (ACCHK_REPORT)
            ctx.obj.conn.action = "INSERT.raw.data"
            query = f"INSERT INTO {ctx.obj.conf['database']['tableraw']}" \
                    f"(year,track,data) " \
                    f"VALUES({line.strip()})"

            # Execute query
            ctx.obj.cur.execute(query)
            click.echo(f"[{step}/{iters}] - {query}")

            # Commit changes
            if step % commit_every == 0:
                # Instrumentation: Set up module action (ACCHK_REPORT)
                ctx.obj.conn.action = "COMMIT.raw.data"
                ctx.obj.conn.commit()
                click.echo(f"[{step}/{iters}] - COMMIT")

            # Wait before the next iteration
            time.sleep(delay)

        # Check the last commit
        if iters % commit_every != 0:
            # Instrumentation: Set up module action (ACCHK_REPORT)
            ctx.obj.conn.action = "COMMIT.raw.data"
            ctx.obj.conn.commit()
            click.echo(f"[{iters}/{iters}] - COMMIT")
    except cx_Oracle.DatabaseError as err:
        click.echo(err)
        ctx.exit(1)
    except KeyboardInterrupt as _:
        click.echo("Error - Interrupted by the user")
        ctx.exit(1)
    finally:
        # Check for the last commit
        if iters % commit_every != 0:
            # Instrumentation: Set up module action (ACCHK_REPORT)
            ctx.obj.conn.action = "COMMIT.raw.data"
            ctx.obj.conn.commit()
            click.echo(f"[{iters}/{iters}] - COMMIT")
